ibc/ibc_handle_event_test.go

```diff
@@ -0,0 +1,242 @@
+	mgr, _, _, _, ibcMod := prepareEnvironment(t, 1, 0, 0, 0)
+	tmpDir := mgr.GetConfig().IBC.StoresDir + "/caches"
+	ibcHost := ibcMod.GetBus().GetIBCHost()
+	store, err := ibcHost.GetProvableStore("test")
+	require.NoError(t, err)
+
+	kvs := []struct {
+		key   []byte
+		value []byte
+	}{
+		{
+			key:   []byte("key1"),
+			value: []byte("value1"),
+		},
+		{
+			key:   []byte("key2"),
+			value: []byte("value2"),
+		},
+		{
+			key:   []byte("key3"),
+			value: []byte("value3"),
+		},
+		{
+			key:   []byte("key4"),
+			value: nil,
+		},
+	}
+
+	testCases := []struct {
+		name            string
+		heights         []uint64
+		expectedHeights []uint64
+		length          int
+	}{
+		{
+			name:            "No pruning after single height increase",
+			heights:         []uint64{1, 2},
+			expectedHeights: []uint64{1},
+			length:          4,
+		},
+		{
+			name:            "No pruning after two height increase",
+			heights:         []uint64{1, 2, 3},
+			expectedHeights: []uint64{1, 2},
+			length:          8,
+		},
+		{
+			name:            "No pruning at max height stored = 3",
+			heights:         []uint64{1, 2, 3, 4},
+			expectedHeights: []uint64{1, 2, 3},
+			length:          12,
+		},
+		{
+			name:            "Pruning after 4 height increase",
+			heights:         []uint64{1, 2, 3, 4, 5},
+			expectedHeights: []uint64{2, 3, 4},
+			length:          12,
+		},
+		{
+			name:            "Pruning after 5 height increase",
+			heights:         []uint64{1, 2, 3, 4, 5, 6},
+			expectedHeights: []uint64{3, 4, 5},
+			length:          12,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// set height
+			publishNewHeightEvent(t, ibcMod.GetBus(), tc.heights[0])
+
+			for _, height := range tc.heights[1:] {
+				for _, kv := range kvs {
+					if kv.value != nil {
+						require.NoError(t, store.Set(kv.key, kv.value))
+					} else {
+						require.NoError(t, store.Delete(kv.key))
+					}
+				}
+
+				// increment the height
+				publishNewHeightEvent(t, ibcMod.GetBus(), height)
+			}
+
+			cache, err := kvstore.NewKVStore(tmpDir)
+			//nolint:errcheck // ignore error just make sure closes incase anything else fails
+			defer cache.Stop()
+			require.NoError(t, err)
+			keys, values, err := cache.GetAll([]byte{}, false)
+			require.NoError(t, err)
+			require.Len(t, keys, tc.length)
+			require.Len(t, values, tc.length)
+
+			// iterate over the expected heights and check the keys and values are correct
+			for i, height := range tc.expectedHeights {
+				for j, key := range keys[i*4 : (i+1)*4] {
+					require.Equal(t, string(key), prepareCacheKey(height, kvs[j%4].key))
+					if kvs[j%4].value == nil {
+						require.Equal(t, values[j], []byte{})
+						continue
+					}
+					require.Equal(t, values[j], kvs[j%4].value)
+				}
+			}
+
+			err = cache.ClearAll()
+			require.NoError(t, err)
+
+			newKeys, newValues, err := cache.GetAll([]byte{}, false)
+			require.NoError(t, err)
+			require.Len(t, newKeys, 0)
```

Comment:
What does this extra check (after clearing everything) provide in the context of this test?